File: Network/Protocol.cpp
#include "Protocol.hpp"

#include <sstream>

#include "io/BoardParser.hpp"

namespace kungfu::net {

namespace {

// Decimal digits of precision for a piece's fractional x/y position in a STATE
// broadcast - enough for smooth interpolation without needlessly verbose lines.
constexpr int kCoordinatePrecisionDigits = 6;

// Single-char color/type codes and 2-char "wR"-style piece tokens, matching the exact
// convention BoardParser/BoardPrinter already use for the text board DSL.
char colorChar(PlayerColor color) { return color == PlayerColor::White ? 'w' : 'b'; }
PlayerColor colorFromChar(char c) { return c == 'w' ? PlayerColor::White : PlayerColor::Black; }

char typeChar(PieceType type) {
    return pieceTypeChar(type);
}

PieceType typeFromChar(char c) {
    switch (c) {
        case 'K': return PieceType::King;
        case 'Q': return PieceType::Queen;
        case 'R': return PieceType::Rook;
        case 'B': return PieceType::Bishop;
        case 'N': return PieceType::Knight;
        default:  return PieceType::Pawn;
    }
}

std::string pieceToken(PlayerColor color, PieceType type) {
    return std::string(1, colorChar(color)) + std::string(1, typeChar(type));
}

std::vector<std::string> splitLines(const std::string& text) {
    std::vector<std::string> lines;
    std::istringstream iss(text);
    std::string line;
    while (std::getline(iss, line)) {
        const std::string trimmed = BoardParser::trim(line);
        if (!trimmed.empty()) {
            lines.push_back(trimmed);
        }
    }
    return lines;
}

}  // namespace

namespace {

char joinModeChar(JoinMode mode) {
    switch (mode) {
        case JoinMode::CreateRoom: return 'C';
        case JoinMode::JoinRoom:   return 'R';
        default:                  return 'Q';
    }
}

}  // namespace

std::string encodeLogin(const std::string& username, const std::string& password) {
    return "LOGIN " + username + " " + password;
}

std::string encodeLoginOk(int rating) {
    return "LOGIN_OK " + std::to_string(rating);
}

std::string encodeLoginFail(const std::string& reason) {
    return "LOGIN_FAIL " + reason;
}

std::string encodeJoin(JoinMode mode, const std::string& room) {
    std::string out = std::string("JOIN ") + joinModeChar(mode);
    if (mode == JoinMode::JoinRoom) {
        out += " " + room;
    }
    return out;
}

std::string encodeClick(int row, int col) {
    return "CLICK " + std::to_string(row) + " " + std::to_string(col);
}

std::string encodeWelcome(PlayerColor color) {
    return std::string("WELCOME ") + colorChar(color);
}

std::string encodeSpectate() {
    return "SPECTATE";
}

std::string encodeRoom(const std::string& key) {
    return "ROOM " + key;
}

std::string encodePlayers(const std::string& white, const std::string& black) {
    return "PLAYERS " + white + " " + black;
}

std::string encodeNoOpponent() {
    return "NO_OPPONENT";
}

std::string encodeForfeitWarning(PlayerColor disconnectedColor, int graceMs) {
    return std::string("FORFEIT_WARNING ") + colorChar(disconnectedColor) + " " + std::to_string(graceMs);
}

std::string encodeForfeit(PlayerColor winner) {
    return std::string("FORFEIT ") + colorChar(winner);
}

std::string encodeRatings(int whiteRating, int blackRating) {
    return "RATINGS " + std::to_string(whiteRating) + " " + std::to_string(blackRating);
}

std::string encodeState(const std::vector<RenderPiece>& pieces) {
    std::ostringstream out;
    out << "STATE\n" << std::fixed;
    out.precision(kCoordinatePrecisionDigits);
    for (const auto& rp : pieces) {
        out << rp.id << " " << pieceToken(static_cast<PlayerColor>(rp.color), static_cast<PieceType>(rp.type))
            << " " << rp.x << " " << rp.y << " " << rp.state << " " << rp.cooldownMs << " "
            << rp.cooldownTotalMs << "\n";
    }
    out << "END";
    return out.str();
}

std::string encodeMoveStarted(const MoveStarted& event) {
    std::ostringstream out;
    out << "MOVE " << pieceToken(event.color, event.type) << " " << event.from.row() << " "
        << event.from.col() << " " << event.to.row() << " " << event.to.col() << " "
        << (event.isCapture ? 1 : 0) << " " << event.elapsedMs << " " << event.notation;
    return out.str();
}

std::string encodePieceCaptured(const PieceCaptured& event) {
    std::ostringstream out;
    out << "CAPTURE " << colorChar(event.capturingColor) << " " << typeChar(event.capturedType) << " "
        << event.at.row() << " " << event.at.col();
    return out.str();
}

std::string encodeGameStarted() {
    return "GAME_START";
}

std::string encodeGameEnded(const GameEnded& event) {
    return std::string("GAME_END ") + colorChar(event.winner);
}

DecodedMessage decode(const std::string& text) {
    const auto lines = splitLines(text);
    if (lines.empty()) {
        return std::monostate{};
    }

    const auto tokens = BoardParser::split(lines[0]);
    if (tokens.empty()) {
        return std::monostate{};
    }
    const std::string& cmd = tokens[0];

    if (cmd == "LOGIN" && tokens.size() >= 3) {
        return LoginMessage{tokens[1], tokens[2]};
    }

    if (cmd == "LOGIN_OK" && tokens.size() >= 2) {
        return LoginOkMessage{std::stoi(tokens[1])};
    }

    if (cmd == "LOGIN_FAIL" && tokens.size() >= 2) {
        return LoginFailMessage{tokens[1]};
    }

    if (cmd == "JOIN" && tokens.size() >= 2 && !tokens[1].empty()) {
        const char modeChar = tokens[1][0];
        if (modeChar == 'R') {
            if (tokens.size() < 3) {
                return std::monostate{};
            }
            return JoinMessage{JoinMode::JoinRoom, tokens[2]};
        }
        const JoinMode mode = modeChar == 'C' ? JoinMode::CreateRoom : JoinMode::QuickMatch;
        return JoinMessage{mode, ""};
    }

    if (cmd == "CLICK" && tokens.size() >= 3) {
        return ClickMessage{std::stoi(tokens[1]), std::stoi(tokens[2])};
    }

    if (cmd == "WELCOME" && tokens.size() >= 2 && !tokens[1].empty()) {
        return WelcomeMessage{colorFromChar(tokens[1][0])};
    }

    if (cmd == "SPECTATE") {
        return SpectateMessage{};
    }

    if (cmd == "ROOM" && tokens.size() >= 2) {
        return RoomMessage{tokens[1]};
    }

    if (cmd == "PLAYERS" && tokens.size() >= 3) {
        return PlayersMessage{tokens[1], tokens[2]};
    }

    if (cmd == "NO_OPPONENT") {
        return NoOpponentMessage{};
    }

    if (cmd == "FORFEIT_WARNING" && tokens.size() >= 3 && !tokens[1].empty()) {
        return ForfeitWarningMessage{colorFromChar(tokens[1][0]), std::stoi(tokens[2])};
    }

    if (cmd == "FORFEIT" && tokens.size() >= 2 && !tokens[1].empty()) {
        return ForfeitMessage{colorFromChar(tokens[1][0])};
    }

    if (cmd == "RATINGS" && tokens.size() >= 3) {
        return RatingsMessage{std::stoi(tokens[1]), std::stoi(tokens[2])};
    }

    if (cmd == "STATE") {
        StateMessage msg;
        for (size_t i = 1; i < lines.size(); ++i) {
            if (lines[i] == "END") {
                break;
            }
            const auto fields = BoardParser::split(lines[i]);
            if (fields.size() < 7 || fields[1].size() < 2) {
                continue;  // malformed line - skip rather than abort the whole snapshot
            }
            RenderPiece rp;
            rp.id = static_cast<uintptr_t>(std::stoull(fields[0]));
            rp.color = static_cast<int>(colorFromChar(fields[1][0]));
            rp.type = static_cast<int>(typeFromChar(fields[1][1]));
            rp.x = std::stod(fields[2]);
            rp.y = std::stod(fields[3]);
            rp.state = std::stoi(fields[4]);
            rp.cooldownMs = std::stod(fields[5]);
            rp.cooldownTotalMs = std::stod(fields[6]);
            msg.pieces.push_back(rp);
        }
        return msg;
    }

    if (cmd == "MOVE" && tokens.size() >= 9 && tokens[1].size() >= 2) {
        MoveStarted event;
        event.color = colorFromChar(tokens[1][0]);
        event.type = typeFromChar(tokens[1][1]);
        event.from = Position(std::stoi(tokens[2]), std::stoi(tokens[3]));
        event.to = Position(std::stoi(tokens[4]), std::stoi(tokens[5]));
        event.isCapture = tokens[6] == "1";
        event.elapsedMs = std::stoi(tokens[7]);
        event.notation = tokens[8];
        return event;
    }

    if (cmd == "CAPTURE" && tokens.size() >= 5 && !tokens[1].empty() && !tokens[2].empty()) {
        PieceCaptured event;
        event.capturingColor = colorFromChar(tokens[1][0]);
        event.capturedType = typeFromChar(tokens[2][0]);
        event.at = Position(std::stoi(tokens[3]), std::stoi(tokens[4]));
        return event;
    }

    if (cmd == "GAME_START") {
        return GameStarted{};
    }

    if (cmd == "GAME_END" && tokens.size() >= 2 && !tokens[1].empty()) {
        return GameEnded{colorFromChar(tokens[1][0])};
    }

    return std::monostate{};
}

}  // namespace kungfu::net

File: WebClient/protocol.js
// JS mirror of Network/Protocol.hpp/.cpp - the wire format lives in exactly one place on the
// C++ side, this is the browser-side equivalent of that same text protocol. Keep in sync with
// Protocol.cpp by hand; there is no shared source of truth across the language boundary.

export const PieceState = {
  Idle: 0,
  Moving: 1,
  Cooldown: 2,
  Airborne: 3,
  ShortRest: 4,
  Captured: 5,
};

const PIECE_GLYPH = {
  K: { w: "♔", b: "♚" },
  Q: { w: "♕", b: "♛" },
  R: { w: "♖", b: "♜" },
  B: { w: "♗", b: "♝" },
  N: { w: "♘", b: "♞" },
  P: { w: "♙", b: "♟" },
};

export function pieceGlyph(color, type) {
  return PIECE_GLYPH[type][color];
}

// Mirrors Logic/history/GameRecord.cpp's pieceValue() - knight/bishop=3, rook=5, queen=9,
// pawn=1, king=0 (capturing a king ends the game before a score would matter).
const PIECE_VALUE = { K: 0, Q: 9, R: 5, B: 3, N: 3, P: 1 };

export function pieceValue(type) {
  return PIECE_VALUE[type] ?? 0;
}

export function encodeLogin(username, password) {
  return `LOGIN ${username} ${password}`;
}

// mode: "Q" | "C" | "R". room only meaningful for "R".
export function encodeJoin(mode, room = "") {
  return mode === "R" ? `JOIN R ${room}` : `JOIN ${mode}`;
}

export function encodeClick(row, col) {
  return `CLICK ${row} ${col}`;
}

function splitTokens(line) {
  return line.trim().split(/\s+/).filter((t) => t.length > 0);
}

function colorFromChar(c) {
  return c === "w" ? "w" : "b";
}

// Returns { type: "<MessageType>", ...fields } for every message Protocol.cpp's decode()
// recognizes, or { type: "unknown" } for anything malformed/unrecognized (mirrors decode()
// returning std::monostate).
export function decode(text) {
  const lines = text.split("\n").map((l) => l.trim()).filter((l) => l.length > 0);
  if (lines.length === 0) {
    return { type: "unknown" };
  }
  const tokens = splitTokens(lines[0]);
  if (tokens.length === 0) {
    return { type: "unknown" };
  }
  const cmd = tokens[0];

  switch (cmd) {
    case "LOGIN_OK":
      if (tokens.length >= 3) {
        return { type: "LoginOk", rating: parseInt(tokens[1], 10), accountCreated: tokens[2] === "1" };
      }
      break;
    case "LOGIN_FAIL":
      if (tokens.length >= 2) {
        return { type: "LoginFail", reason: tokens[1] };
      }
      break;
    case "WELCOME":
      if (tokens.length >= 2 && tokens[1]) {
        return { type: "Welcome", color: colorFromChar(tokens[1][0]) };
      }
      break;
    case "SPECTATE":
      return { type: "Spectate" };
    case "ROOM":
      if (tokens.length >= 2) {
        return { type: "Room", key: tokens[1] };
      }
      break;
    case "PLAYERS":
      if (tokens.length >= 3) {
        return { type: "Players", white: tokens[1], black: tokens[2] };
      }
      break;
    case "NO_OPPONENT":
      return { type: "NoOpponent" };
    case "FORFEIT_WARNING":
      if (tokens.length >= 3 && tokens[1]) {
        return { type: "ForfeitWarning", color: colorFromChar(tokens[1][0]), graceMs: parseInt(tokens[2], 10) };
      }
      break;
    case "FORFEIT":
      if (tokens.length >= 2 && tokens[1]) {
        return { type: "Forfeit", winner: colorFromChar(tokens[1][0]) };
      }
      break;
    case "RECONNECTED":
      if (tokens.length >= 2 && tokens[1]) {
        return { type: "Reconnected", color: colorFromChar(tokens[1][0]) };
      }
      break;
    case "RATINGS":
      if (tokens.length >= 3) {
        return { type: "Ratings", white: parseInt(tokens[1], 10), black: parseInt(tokens[2], 10) };
      }
      break;
    case "STATE": {
      const pieces = [];
      for (let i = 1; i < lines.length; i++) {
        if (lines[i] === "END") break;
        const fields = splitTokens(lines[i]);
        if (fields.length < 7 || fields[1].length < 2) continue;
        pieces.push({
          id: fields[0],
          color: colorFromChar(fields[1][0]),
          type: fields[1][1],
          x: parseFloat(fields[2]),
          y: parseFloat(fields[3]),
          state: parseInt(fields[4], 10),
          cooldownMs: parseFloat(fields[5]),
          cooldownTotalMs: parseFloat(fields[6]),
        });
      }
      return { type: "State", pieces };
    }
    case "MOVE":
      if (tokens.length >= 9 && tokens[1] && tokens[1].length >= 2) {
        return {
          type: "Move",
          color: colorFromChar(tokens[1][0]),
          pieceType: tokens[1][1],
          fromRow: parseInt(tokens[2], 10),
          fromCol: parseInt(tokens[3], 10),
          toRow: parseInt(tokens[4], 10),
          toCol: parseInt(tokens[5], 10),
          isCapture: tokens[6] === "1",
          elapsedMs: parseInt(tokens[7], 10),
          notation: tokens[8],
        };
      }
      break;
    case "CAPTURE":
      if (tokens.length >= 5 && tokens[1] && tokens[2]) {
        return {
          type: "Capture",
          capturingColor: colorFromChar(tokens[1][0]),
          capturedType: tokens[2][0],
          row: parseInt(tokens[3], 10),
          col: parseInt(tokens[4], 10),
        };
      }
      break;
    case "GAME_START":
      return { type: "GameStarted" };
    case "GAME_END":
      if (tokens.length >= 2 && tokens[1]) {
        return { type: "GameEnded", winner: colorFromChar(tokens[1][0]) };
      }
      break;
  }
  return { type: "unknown" };
}
